### ISSUE TITLE
Warn for unsafe variables in expressions, have consistent behavior for interpreted and compiled code

### ISSUE THREAD

**Issue by mheiber** [erlang/otp#5379]
**Describe the bug**

erl_lint warns when a variable is bound in one branch of a pattern match but not another (unsafe variable).
However, there is no warning when a variable is bound in one branch of an operator expression but not another.
This is a problem because:
- Makes it so we cannot determine statically whether a variable is bound
- Leads to distinct behavior for interpreted vs. compiled code for something one would expect to be deterministic (variable binding)

**To Reproduce**

Save the following as `sample.erl`:

```erl
-module(sample).
-export([main/1]).

main(_) ->
    (begin X=1, true end)
        and
    (begin F=(fun () -> X=2 end), F(), true end),
    io:format("~p~n", [X]).
```

`escript sample.erl` warns that 'X' is unused, then uses `X`:

```
sample.erl:6:25: Warning: variable 'X' is unused
1
```

`escript -c sample.erl` (compile+run) blows up:

```
escript: exception error: no match of right hand side value 2
```

**Expected behavior**

- Warning or error for unsafe variable binding
- Consistent behavior for interpreted and compiled code

**Affected versions**

Erlang/OTP 25 [DEVELOPMENT] [erts-12.0.4] [source-5979a8621b] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns] [dtrace]

Probably earlier versions as well

Credit @rgrig @hajduakos for finding this issue

**Comment by rgrig** [comment]
One way to think about "unsafe variables" is that, when you reach an occurrence X, you should be able to determine statically if X is bound or not; in other words, for *all* execution paths that reach this occurrence, either all bind X or neither binds X. The manual says things like "it doesn't matter which branch of a `case` you take", which can be viewed as a special case of the more general rule. The issue is that it is not only `if`, `case` and `receive` that introduce branching. It's also the evaluation order in expressions.

So, if X occurs in expression E1+E2 (or some other operator other than + that does not guarantee order of evaluation) and X is not bound when E1+E2 is reached, then X should occur at most once in E1+E1 (in the lhs of a match). If it occurs in a non-binding position it is an error; if it occurs in a binding position more than once then when you reach one of those bindings you either (a) executed the other one or (b) you did not. So, it's fair to regard it as "unsafe".

(BTW, what I find especially bothering in the example above is that the variables captured by the lambda cannot be determined statically: it depends on which side of the `and` gets executed first.)

**Comment by bjorng** [comment]
I don't see any reason to leave the evaluation order undefined. Historically, I think were some differences in evaluation order between JAM and BEAM, and I think we changed BEAM to evaluate left to right, which JAM already did. I suggest that we document that the evaluation order is left to right and that we change `erl_eval` to evaluate left to right. The OTP Technical Board will have to approve this decision.

**Comment by bjorng** [comment]
@rgrig I would say that in the example, `X` is safe (in the sense that it is always bound) in the call to `io:format/2` regardless of evaluation order. The reason is that if `io:format/2` is reached, both sides of `and` must have been successfully evaluated and `X` is guaranteed to have been bound.

Another way to reconcile the compiler and `erl_eval` would be to change the compiler to behave as `erl_eval`. That essentially means that when evaluating the left and right sides of an operators, variables bound in the lefthand side will not be seen in the righthand side and vice versa. For this particular example it would mean the code would be essentially be transformed like this:

```
(begin X = 1, true end)
        and
    (begin F=(fun () -> X1 = 2 end), F(), true end),
    io:format("~p~n", [X]).
```

If `X` is bound on both sides of the operator like in:

```
main(_) ->
    (begin X = 1, true end)
        and
    (begin X = 2, true end),
    io:format("~p~n", [X]).
```

the code would be transformed like this:

```
main(_) ->
    (begin X1 = 1, true end)
        and
    (begin X2 = 2, true end),
    X = case X1 =:= X2 of
             true -> X1;
             false -> error({badmatch,X2})
    end,          
    io:format("~p~n", [X]).
```

In practice, after simplifications and optimizations, the same code will be generated as the current compiler generates.

**Comment by rgrig** [comment]
Yes it is safe at `io:format`, but my argument was whether it is safe when you reach one of the branches. When you reach right-branch you may or may have not seen the left branch -> hence unsafe.

Both your proposals would fix this. And by your proposals I mean:
(a) fix left-to-right order and say that you see what was to the left; or
(b) explicitly say that you don't see what is in other branches, but then you need the check at the end.

**Comment by bjorng** [comment]
I was going to summarize my proposals, but you have already done it.

> Yes it is safe at `io:format`, but my argument was whether it is safe when you reach one of the branches. When you reach right-branch you may or may have not seen the left branch -> hence unsafe.

Good point. It turns out that if I try to use the value of a variable defined on the lefthand side on the righthand side, it is always an error:

```
main(_) ->
    (begin X = 1, true end)
        and
    (begin X+2, true end),
    io:format("~p~n", [X]).
```

```
t.erl:10:12: variable 'X' is unbound
%   10|     (begin X+2, true end),
%     |            ^
```

Given that, I think that second proposal (b) make most sense.

**Comment by bjorng** [comment]
Bug fix in this PR: #5456
